### PR TITLE
Added depends_on attribute

### PR DESCRIPTION
**Pull request by eko**
A Homebrew formula allows to specify one or multiple formulas that are
needed in order to build the current one.

This is done by the 'depends_on' attribute: https://docs.brew.sh/Formula-Cookbook#specifying-other-formulae-as-dependencies

In this pull request I've added this attribute.

I would be happy to have it merge as I need a dependency in my case ;-)

Thank you

### CODE REVIEW

**Comment by eko** [comment]
Hi @Justintime50,

Are you interested by this addition?

Thank you

**Comment by Justintime50** [comment]
Hey @eko, thanks for the contribution! This looks pretty good and I'm interested in getting it included.

I've been out for the last few days. I will take another look at this over the next few days and get back to you about getting it merged.

File: test/unit/test_formula.py [hunk 1]
```diff
@@ -38,17 +42,45 @@ def test_generate_formula():
         # - punctuation
         # - trailing whitespace
         # - extra capitilization
-        'description': 'A tool to release... scripts, binaries, and executables to GitHub. ',
```
Comment (Justintime50) [inline test/unit/test_formula.py:41]: Please leave the description as it was. If you look at the comment in the code directly above it, it explains why it was written the way it was:

>         # We use a badly written description string here on purpose to test our formatting code, this includes...

File: test/cassettes/test_formula_template.rb [hunk 2]
```diff
@@ -3,12 +3,15 @@
 
 # This file was generated by Homebrew Releaser. DO NOT EDIT.
 class HomebrewReleaser < Formula
-  desc "Tool to release scripts, binaries, and executables to github"
```
Comment (Justintime50) [inline test/cassettes/test_formula_template.rb:6]: Once the description is returned to how it shows up in the test, this field should not be altered in this PR. (see notes below)

File: test/unit/test_formula.py [hunk 3]
```diff
@@ -38,17 +42,45 @@ def test_generate_formula():
...
+def test_generate_formula_no_depends_on():
+    """Tests that we generate the formula content correctly (when no depends_on given).
+
+    This test generates the formula "cassette" file if it does not exist already
+    To regenerate the file (when changes are made to how formula are generated), simply
+    delete the file and run tests again. Ensure the output of the file is correct.
+    """
+    repository = {
+        # Here we don't start the description off with an article
+        'description': 'Release scripts, binaries, and executables to GitHub',
```
Comment (Justintime50) [inline test/unit/test_formula.py:73]: For this test, the description field and comment here about starting with an article aren't important since this is covered in other tests.

File: test/unit/test_formula.py [hunk 4]
```diff
@@ -38,17 +42,45 @@ def test_generate_formula():
...
+        TAR_URL,
+        DEPENDS_ON,
+        TEST,
+    )
     cassette_filename = 'test_formula_template.rb'
 
     _record_cassette(CASSETTE_PATH, cassette_filename, formula)
 
 
-def test_generate_formula_no_article_description():
```
Comment (Justintime50) [inline test/unit/test_formula.py:51]: Please leave this test intact and simply add a new test to verify your "depend_on" logic works. Replacing tests could lead to regressions if the test assertions got changed or could lead to confusion if we are testing two things inside of one test and eventually those assertions could get lost.

File: test/unit/test_formula.py [hunk 5]
```diff
@@ -38,17 +42,45 @@ def test_generate_formula():
...
+        'license': {'spdx_id': 'MIT'},
+    }
+
+    formula = Formula.generate_formula_data(USERNAME, REPO_NAME, repository, CHECKSUM, INSTALL, TAR_URL)
+    cassette_filename = 'test_formula_template_no_depends_on.rb'
+
+    _record_cassette(CASSETTE_PATH, cassette_filename, formula)
+
+
+def test_generate_formula_no_test():
```
Comment (Justintime50) [inline test/unit/test_formula.py:83]: Thanks for adding the "no test in formula" test here! After looking at it though, it's not correct (the docstring mentions no article in the description field which is a different test entirely). Please update the docstring to be accurate for the test in question.

**Review by Justintime50 — CHANGES_REQUESTED** [review]
Overall this looks pretty good! I have a few suggestions to clean up an item or two.

The biggest things that need corrections are the tests which appear to be altered in a way that could lead to regressions and confusion in the future. Let me know if you have questions. 

Once these have been tweaked, I'll take another look and we can get it merged and released once everything looks good.

**Comment by eko** [comment]
Hi @Justintime50, thank you for your comments! I misunderstood what was done on the project, it's quite clear now.

I've fixed your comments, don't hesitate if it still some fixes to do.

File: homebrew_releaser/formula.py [hunk 6]
```diff
@@ -50,6 +51,14 @@ def generate_formula_data(
         if first_word_of_desc[0].lower() in articles:
             description = first_word_of_desc[1].strip().capitalize()
 
+        depends_on_content = ''
+
+        if depends_on:
+            dependencies = [dependency.strip() for dependency in depends_on.split('\n') if dependency]
```
Comment (Justintime50) [inline homebrew_releaser/formula.py:57]: I'm not sure the `if dependency` is needed here since if there is nothing present, it'll be an empty list anyway and iterating an empty list won't do anything. Please let me know though if I've missed something or misunderstood.

```suggestion
            dependencies = [dependency.strip() for dependency in depends_on.split('\n')]
```

File: test/unit/test_formula.py [hunk 7]
```diff
@@ -63,6 +76,42 @@ def test_generate_formula_no_article_description():
     }
 
     formula = Formula.generate_formula_data(USERNAME, REPO_NAME, repository, CHECKSUM, INSTALL, TAR_URL)
-    cassette_filename = 'test_formula_template_no_article_description.rb'
+    cassette_filename = 'test_formula_template_no_depends_on.rb'
```
Comment (Justintime50) [inline test/unit/test_formula.py:79]: The `cassette_filename` should match the test name, in this case it should be `test_formula_template_no_article_description`

File: test/unit/test_formula.py [hunk 8]
```diff
@@ -42,7 +46,16 @@ def test_generate_formula():
...
-    formula = Formula.generate_formula_data(USERNAME, REPO_NAME, repository, CHECKSUM, INSTALL, TAR_URL, TEST)
+    formula = Formula.generate_formula_data(
+        USERNAME,
+        REPO_NAME,
+        repository,
+        CHECKSUM,
+        INSTALL,
+        TAR_URL,
+        DEPENDS_ON,
+        TEST,
```
Comment (Justintime50) [inline test/unit/test_formula.py:57]: Because we are adding more parameters, can we please update line 33 to say `Tests that we generate the formula content correctly when all parameters are passed.` This will help clarify that this test tests **everything**, following tests will test when certain parameters are missing and that they are handled gracefully.

File: test/unit/test_formula.py [hunk 9]
```diff
@@ -63,6 +76,42 @@ def test_generate_formula_no_article_description():
...
+    This test generates the formula "cassette" file if it does not exist already
+    To regenerate the file (when changes are made to how formula are generated), simply
+    delete the file and run tests again. Ensure the output of the file is correct.
+    """
+    repository = {
+        'description': 'Release scripts, binaries, and executables to GitHub',
+        'license': {'spdx_id': 'MIT'},
+    }
+
+    formula = Formula.generate_formula_data(USERNAME, REPO_NAME, repository, CHECKSUM, INSTALL, TAR_URL)
```
Comment (Justintime50) [inline test/unit/test_formula.py:96]: Can we explicitly include test here and `None` for the depends on?
```suggestion
    formula = Formula.generate_formula_data(USERNAME, REPO_NAME, repository, CHECKSUM, INSTALL, TAR_URL, None, TEST)
```

**Review by Justintime50 — CHANGES_REQUESTED** [review]
Thanks for the updates! There are still a couple minor things that don't line up. Once these are addressed, I'll be good to merge.

File: homebrew_releaser/formula.py (hunk 6 above)
Comment (eko) [reply homebrew_releaser/formula.py]: This was because the string was not stripped, I updated the condition to simplify it by calling `depends_on.strip().split('\n')`.

Before that `depends_on.split('\n')` was returning: `['', '"gcc"', '"bash" => :build', '']`

File: test/unit/test_formula.py (hunk 7 above)
Comment (eko) [reply test/unit/test_formula.py]: Of course, sorry for this mistake :)

**Comment by eko** [comment]
Thanks, I just updated the PR with your latest comments

File: homebrew_releaser/formula.py (hunk 6 above)
Comment (eko) [reply homebrew_releaser/formula.py]: This is due to the variable declaration at the top of the `formula_test` file:

```python
DEPENDS_ON = """
"gcc"
"bash" => :build
"""
```

Maybe you have a tip in Python to avoid new lines ? I'm a newcomer to Python ;-)

Comment (Justintime50) [reply homebrew_releaser/formula.py]: Ah that makes sense, in that case, you'll definitely want to put this line back to `dependencies = [dependency.strip() for dependency in depends_on.split('\n') if dependency]`. What you changed it to won't work as it's going to strip the whole string and not each individual piece after splitting on the newline.

File: homebrew_releaser/formula.py [hunk 10]
```diff
@@ -50,6 +51,13 @@ def generate_formula_data(
         if first_word_of_desc[0].lower() in articles:
             description = first_word_of_desc[1].strip().capitalize()
 
+        depends_on_content = ''
+
+        if depends_on:
+            for dependency in depends_on.strip().split('\n'):
```
Comment (Justintime50) [inline homebrew_releaser/formula.py:57]: You'll definitely want to put this back to `dependencies = [dependency.strip() for dependency in depends_on.split('\n') if dependency]` per the convo on the PR elsewhere.

**Comment by Justintime50** [comment]
Tests look good now, thanks! You'll just want to revert the one line I left a comment on and this will be ready!

File: homebrew_releaser/formula.py (hunk 10 above)
Comment (Justintime50) [inline homebrew_releaser/formula.py:57]:
```suggestion
            dependencies = [dependency.strip() for dependency in depends_on.split('\n') if dependency]

            for dependency in dependencies:
```

**Review by Justintime50 — CHANGES_REQUESTED** [review]

File: homebrew_releaser/formula.py (hunk 10 above)
Comment (eko) [reply homebrew_releaser/formula.py]: This is done, thank you :)

**Review by Justintime50 — APPROVED** [review]
Thanks for addressing all the feedback! This is now ready to merge.

Thanks again for your contribution!

**Comment by eko** [comment]
Cool, thank you for your time!

**Comment by Justintime50** [comment]
This was just released, feel free to give it a shot!
